src/badge/examples/EthereumYearBadge.sol

```diff
@@ -0,0 +1,135 @@
+    error ErrorInvalidSignature();
+
+    /// @notice Emitted when the offchain signer is updated.
+    /// @param oldSigner The address of previous offchain signer.
+    /// @param newSigner The address of current offchain signer.
+    event UpdateOffchainSigner(address indexed oldSigner, address indexed newSigner);
+
+    /// @dev The type hash for issue badge.
+    // solhint-disable-next-line var-name-mixedcase
+    bytes32 private constant _ISSUE_BADGE_TYPEHASH = keccak256("IssueBadge(address recipient,uint256 year)");
+
+    /// @notice The address of badge offchain signer.
+    address public offchainSigner;
+
+    /// @notice The base token URI.
+    string public baseTokenURI;
+
+    // badge UID => current year
+    mapping(bytes32 => uint256) public badgeYear;
+
+    constructor(address resolver_, address offchainSigner_, string memory baseTokenURI_)
+        ScrollBadge(resolver_)
+        EIP712("Ethereum Year Badge", "1")
+    {
+        _updateOffchainSigner(offchainSigner_);
+
+        baseTokenURI = baseTokenURI_;
+    }
+
+    /// @notice Update the offchain signer.
+    /// @param newSigner The new offchain signer address.
+    function updateOffchainSigner(address newSigner) external onlyOwner {
+        _updateOffchainSigner(newSigner);
+    }
+
+    /// @notice Update the base token URI.
+    /// @param baseTokenURI_ The new base token URI.
+    function updateBaseTokenURI(string memory baseTokenURI_) external onlyOwner {
+        baseTokenURI = baseTokenURI_;
+    }
+
+    /// @inheritdoc ScrollBadge
+    function onIssueBadge(Attestation calldata attestation)
+        internal
+        override (
+            ScrollBadgeSelfAttest,
+            ScrollBadgeCustomPayload,
+            ScrollBadgeNoExpiry,
+            ScrollBadgeNonRevocable,
+            ScrollBadgeSingleton
+        )
+        returns (bool)
+    {
+        if (!super.onIssueBadge(attestation)) {
+            return false;
+        }
+
+        // check signature
+        bytes memory payload = getPayload(attestation);
+        (uint256 year, bytes memory signature) = decodePayloadData(payload);
+        bytes32 structHash = keccak256(abi.encode(_ISSUE_BADGE_TYPEHASH, attestation.recipient, year));
+        bytes32 hash = _hashTypedDataV4(structHash);
+        address signer = ECDSA.recover(hash, signature);
+        if (signer != offchainSigner) revert ErrorInvalidSignature();
+
+        badgeYear[attestation.uid] = year;
+
+        return true;
+    }
+
+    /// @inheritdoc ScrollBadge
+    function onRevokeBadge(Attestation calldata attestation)
+        internal
+        override (ScrollBadge, ScrollBadgeSelfAttest, ScrollBadgeCustomPayload, ScrollBadgeNoExpiry, ScrollBadgeSingleton)
+        returns (bool)
+    {
+        return super.onRevokeBadge(attestation);
+    }
+
+    /// @inheritdoc ScrollBadge
+    function badgeTokenURI(bytes32 uid) public view override returns (string memory) {
+        uint256 year = badgeYear[uid];
```

Comment:
Does this work as intended for:
1. Non-existent uid
2. If uid is `0x` (default token URI)?

Reply:
it's only for frontend, so maybe better to revert on non-existent uid.